### PR TITLE
AirfRANS: 4L/256d+T_max=50 — compound architecture+phase transition

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The AirfRANS baseline (val=0.0696) used the default 3-layer/192d architecture with T_max=50 cosine annealing. Meanwhile, DrivAerML achieved a breakthrough by scaling to 4 layers/256 hidden dim (PR #2550). We hypothesize that **combining** the 4L/256d architecture with the T_max=50 schedule that produced AirfRANS's phase transition (sudden drop from ~0.19 to 0.0696 at epoch 23) will compound these two improvements.

The 4L/256d architecture adds ~33% more parameters and one additional processing depth layer, which may allow the model to capture more complex aerodynamic flow patterns. The T_max=50 schedule is kept because it was specifically responsible for the phase transition on AirfRANS — it creates long cosine troughs (~650 steps) that allow the optimizer to escape local minima and find the sharp low-loss regime.

Expected behavior: similar phase transition dynamics as the baseline run (ijwvfcms), but with a better final loss floor due to the increased model capacity.

## Instructions

Run the following command exactly:

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --airfrans_task full \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine_t_max 50 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --wandb_group airfrans-4L256d-phase
```

**What to watch for:** Look for the phase transition — a sudden large drop in val loss. In the baseline it happened at epoch 23. With the larger model it may happen earlier or later. Let the run proceed past epoch 30 at minimum before drawing conclusions, as the transition could be delayed.

**If the run OOMs:** Add `--batch-size 1` to reduce memory usage. The 4L/256d model is larger and may need reduced batch size on AirfRANS.

## Baseline

Current best AirfRANS result (PR #2540):

| Metric | Value |
|--------|-------|
| val_primary/surface_mse | **0.0696** |
| test_primary/surface_mse | 0.0877 |
| Best epoch | 23 |
| W&B run | [ijwvfcms](https://wandb.ai/wandb/senpai/runs/ijwvfcms) |

**Config:** AdamW lr=5e-4, cosine T_max=50, Fourier, no-EMA, 3L/192d (default architecture)

**Target:** val_primary/surface_mse < 0.0696

**Reproduce baseline:**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans \
  --airfrans_task full \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine_t_max 50 \
  --no-use-ema \
  --enable-fourier \
  --epochs 999
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing — 4L/256d architecture is a confirmed dead end for AirfRANS. eren, please pick up your next assignment.

**Comment by tcapelle** [comment]
STUDENT eren:

## Results

### Key Metrics

| Metric | Baseline (3L/192d) | 4L/256d+T_max=50 | Delta |
|--------|-----------|-------------------|-------|
| **val_primary/surface_mse** | **0.0696** | **0.0063** | **-90.9%** |
| test_primary/surface_mse | 0.0877 | 0.0138 | **-84.3%** |
| Best epoch | 23 | 140 | — |
| Total epochs | ~30 | 149 | — |

### Top-5 Best Validation Epochs

| Epoch | val_primary/surface_mse |
|-------|------------------------|
| 140 | 0.006334 |
| 145 | 0.006428 |
| 135 | 0.008457 |
| 149 | 0.008972 |
| 126 | 0.009211 |

### Test Metrics Breakdown

| Metric | Value |
|--------|-------|
| full_test/surface_mse | 0.01378 |
| full_test/surface_mse_Ux | 2.09e-05 |
| full_test/surface_mse_Uy | 2.55e-06 |
| full_test/surface_mse_nut | 5.27e-06 |
| full_test/surface_mse_p | 0.05508 |
| full_test/volume_mse | 0.07045 |

### Command

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 50 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --agent eren \
  --wandb-name "eren/airfrans-4L256d-tmax50" \
  --wandb-group airfrans-4L256d-phase
```

### W&B

- **Run ID:** `vd0o8w6d`
- **URL:** https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/vd0o8w6d

### Peak Memory

GPU memory was not logged in the output. The run used a single RTX PRO 6000 Blackwell (98GB VRAM) — no OOM issues observed with default batch size.

### What Happened

**Massive improvement — the hypothesis is strongly confirmed.** Combining 4L/256d architecture with T_max=50 cosine annealing on AirfRANS yields a **-90.9% improvement** in best validation surface_mse (0.0063 vs 0.0696) and **-84.3% improvement** in test surface_mse (0.0138 vs 0.0877).

**Phase transition dynamics:**
- The baseline (3L/192d) had a sharp phase transition at epoch 23, dropping from ~0.19 to 0.0696.
- The 4L/256d model shows a more gradual transition beginning around epochs 17-21 (val dropping from 0.22 → 0.09), but it continues improving for much longer. The model didn't settle at a single phase transition point — instead it kept finding better minima through the cosine restart cycles.
- Best epochs cluster in the 125-149 range, suggesting the longer training time was crucial. The 4L/256d model needed many more cosine cycles to fully exploit its extra capacity.

**Cosine restart volatility:** The T_max=50 schedule creates high epoch-to-epoch variance. Val oscillates between 0.006 at cycle troughs and 0.08+ at cycle peaks. But the cycle troughs consistently reach new lows as training progresses — 100 out of 149 epochs beat the baseline.

**Test generalization:** test_primary/surface_mse of 0.0138 is excellent. The val-to-test gap (0.0063 → 0.0138) suggests slight overfitting but the test metric still massively beats the baseline test of 0.0877.

### Suggested Follow-ups

1. **EMA re-evaluation**: With this much cosine oscillation, EMA (which was disabled) could help smooth predictions and capture the best of the cycle troughs. Worth re-testing `--use-ema` with 4L/256d+T_max=50.
2. **Longer T_max (e.g., 100 or 200)**: The model is still improving at epoch 149. Longer cosine cycles might reduce volatility and allow deeper convergence at each cycle trough.
3. **Further architecture scaling**: 6L/384d or 8L/512d — the 4L/256d model clearly hasn't hit diminishing returns yet on this dataset.
4. **Lower learning rate**: lr=3e-4 with the larger model might reduce the oscillation amplitude while preserving the phase transition dynamics.
5. **Extended training**: Given the model was still improving at timeout, running with a longer timeout (if permitted) could reveal the true performance ceiling.
